cfscrape/__init__.py

```diff
@@ -119,27 +120,39 @@ def is_cloudflare_captcha_challenge(resp):
     def request(self, method, url, *args, **kwargs):
         resp = super(CloudflareScraper, self).request(method, url, *args, **kwargs)
 
-        # Check if Cloudflare captcha challenge is presented
-        if self.is_cloudflare_captcha_challenge(resp):
-            self.handle_captcha_challenge(resp, url)
-
         # Check if Cloudflare anti-bot "I'm Under Attack Mode" is enabled
         if self.is_cloudflare_iuam_challenge(resp):
             resp = self.solve_cf_challenge(resp, **kwargs)
 
         return resp
 
+    def send(self, request, **kwargs):
+        resp = super(CloudflareScraper, self).send(request, **kwargs)
+
+        while self.is_cloudflare_captcha_challenge(resp):
+            hook_resp = dispatch_hook('captcha', self.hooks, resp, **kwargs)
+
+            # The captcha hooks are expected to return a different response
+            if resp is hook_resp or not hook_resp:
+                # No captcha hooks or the response is invalid
+                self.handle_captcha_challenge(resp)
+
+            # Replace the response and check again
+            resp = hook_resp
+
+        return resp
+
     def cloudflare_is_bypassed(self, url, resp=None):
         cookie_domain = ".{}".format(urlparse(url).netloc)
         return (
             self.cookies.get("cf_clearance", None, domain=cookie_domain) or
             (resp and resp.cookies.get("cf_clearance", None, domain=cookie_domain))
         )
 
-    def handle_captcha_challenge(self, resp, url):
```

Comment:
It doesn't really make sense to accept the url argument here and it's not documented but we might want to wait until the next major release to remove it.